### PR TITLE
Added some unit tests for the scaledjob_controller.go

### LINKED ISSUE #3661: Extend e2e test coverage for ScaledJob
### Proposal

Cover advanced options/scaling strategies of ScaledJob in e2e tests.

### Use-Case

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by geoffrey1330**
Signed-off-by: geoffrey1330 <israelgeoffrey13@gmail.com>

_Provide a description of what has been changed_

### Checklist

- [ ] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [ ] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [ ] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [ ] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes #3661

Relates to #2357

### CODE REVIEW

File: controllers/keda/scaledjob_controller_test.go [hunk 1]
```diff
@@ -22,6 +23,95 @@ var _ = Describe("ScaledJobController", func() {
...
+			pollingInterval := int32(5)
+			sj.Spec.PollingInterval = &pollingInterval
+			err := k8sClient.Create(context.Background(), sj)
+			Expect(err).ToNot(HaveOccurred())
+
+			// Check if ScaledJob is created successfully
+			Eventually(func() bool {
+				err := k8sClient.Get(context.Background(), types.NamespacedName{Name: sjName, Namespace: "default"}, sj)
+				return err == nil
+			}, 2*time.Minute, 10*time.Second).Should(BeTrue())
```
Comment (JorTurFer) [inline controllers/keda/scaledjob_controller_test.go:59]: I'd reduce this timeout. 2 min for a unit tests looks as too much

File: controllers/keda/scaledjob_controller_test.go [hunk 2]
```diff
@@ -22,6 +23,95 @@ var _ = Describe("ScaledJobController", func() {
...
+			pollingInterval := int32(5)
+			sj.Spec.PollingInterval = &pollingInterval
+			err := k8sClient.Create(context.Background(), sj)
+			Expect(err).ToNot(HaveOccurred())
+
+			// Wait for ScaledJob creation
+			Eventually(func() bool {
+				err := k8sClient.Get(context.Background(), types.NamespacedName{Name: sjName, Namespace: "default"}, sj)
+				return err == nil
+			}, 2*time.Minute, 10*time.Second).Should(BeTrue())
```
Comment (JorTurFer) [inline controllers/keda/scaledjob_controller_test.go:96]: Same as above

File: controllers/keda/scaledjob_controller_test.go [hunk 3]
```diff
@@ -22,6 +23,95 @@ var _ = Describe("ScaledJobController", func() {
...
+
+			// Delete the ScaledJob
+			err = k8sClient.Delete(context.Background(), sj)
+			Expect(err).ToNot(HaveOccurred())
+
+			// Wait for ScaledJob deletion
+			Eventually(func() bool {
+				err := k8sClient.Get(context.Background(), types.NamespacedName{Name: sjName, Namespace: "default"}, sj)
+				return err != nil
+			}, 2*time.Minute, 10*time.Second).Should(BeTrue())
```
Comment (JorTurFer) [inline controllers/keda/scaledjob_controller_test.go:106]: Same as above

File: controllers/keda/scaledjob_controller_test.go [hunk 4]
```diff
@@ -22,6 +23,95 @@ var _ = Describe("ScaledJobController", func() {
...
+
+			// Verify associated resources are deleted
+			jobList := &batchv1.JobList{}
+			err = k8sClient.List(context.Background(), jobList, &client.ListOptions{Namespace: "default"})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(len(jobList.Items)).To(BeZero())
```
Comment (JorTurFer) [inline controllers/keda/scaledjob_controller_test.go:112]: We aren't checking if the jobs were created before checking if they've been removed, so they could not be created and this test won't fail without checking if it works or not. Could you add a validation before the removal to ensure that there are jobs?

**Review by JorTurFer — COMMENTED** [review]
Look good! Thanks for the improvement ❤️ 

Could you add more validations to cover the typical errors? You can check what's tested for ScaledObject controller: https://github.com/kedacore/keda/blob/main/controllers/keda/scaledobject_controller_test.go

File: controllers/keda/scaledjob_controller_test.go (hunk 1 above)
Comment (geoffrey1330) [reply controllers/keda/scaledjob_controller_test.go:59]: Alright.

File: controllers/keda/scaledjob_controller_test.go (hunk 4 above)
Comment (geoffrey1330) [reply controllers/keda/scaledjob_controller_test.go:112]: Alright

File: config/crd/bases/eventing.keda.sh_cloudeventsources.yaml [hunk 5]
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.14.0
+    controller-gen.kubebuilder.io/version: v0.13.0
```
Comment (JorTurFer) [inline config/crd/bases/eventing.keda.sh_cloudeventsources.yaml:6]: Please revert this file changes

File: config/crd/bases/keda.sh_clustertriggerauthentications.yaml [hunk 6]
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.14.0
```
Comment (JorTurFer) [inline config/crd/bases/keda.sh_clustertriggerauthentications.yaml:6]: Please revert this file changes

File: config/crd/bases/keda.sh_scaledobjects.yaml [hunk 7]
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.14.0
```
Comment (JorTurFer) [inline config/crd/bases/keda.sh_scaledobjects.yaml:6]: Please revert this file changes

File: config/crd/bases/keda.sh_triggerauthentications.yaml [hunk 8]
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.14.0
+    controller-gen.kubebuilder.io/version: v0.13.0
```
Comment (JorTurFer) [inline config/crd/bases/keda.sh_triggerauthentications.yaml:6]: Please revert this file changes

File: config/metrics-server/kustomization.yaml [hunk 9]
```diff
@@ -10,5 +10,5 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 images:
 - name: ghcr.io/kedacore/keda-metrics-apiserver
-  newName: ghcr.io/kedacore/keda-metrics-apiserver
```
Comment (JorTurFer) [inline config/metrics-server/kustomization.yaml:13]: Please revert this file changes

File: config/webhooks/kustomization.yaml [hunk 10]
```diff
@@ -7,5 +7,5 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 images:
 - name: ghcr.io/kedacore/keda-admission-webhooks
-  newName: ghcr.io/kedacore/keda-admission-webhooks
```
Comment (JorTurFer) [inline config/webhooks/kustomization.yaml:10]: Please revert this file changes

File: config/manager/kustomization.yaml [hunk 11]
```diff
@@ -6,5 +6,5 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 images:
 - name: ghcr.io/kedacore/keda
-  newName: ghcr.io/kedacore/keda
```
Comment (JorTurFer) [inline config/manager/kustomization.yaml:9]: Please revert this file changes

File: config/metrics-server/kustomization.yaml (hunk 9 above)
Comment (geoffrey1330) [reply config/metrics-server/kustomization.yaml:13]: I forget to revert the changes after installing the operator on my local cluster

File: config/crd/bases/keda.sh_clustertriggerauthentications.yaml (hunk 6 above)
Comment (geoffrey1330) [reply config/crd/bases/keda.sh_clustertriggerauthentications.yaml]: Done

File: config/crd/bases/keda.sh_triggerauthentications.yaml (hunk 8 above)
Comment (geoffrey1330) [reply config/crd/bases/keda.sh_triggerauthentications.yaml]: Done

File: config/webhooks/kustomization.yaml (hunk 10 above)
Comment (geoffrey1330) [reply config/webhooks/kustomization.yaml]: Done

**Comment by JorTurFer** [comment]
The new test is failing with with error:  `[FAILED] Expected jobs to be created`

**Comment by tomkerkhove** [comment]
@geoffrey1330 Any update on this PR?

**Comment by geoffrey1330** [comment]
> @geoffrey1330 Any update on this PR?

Nothing yet.

**Pull request closed without merging**